soroban-sdk/src/env.rs

```diff
@@ -737,8 +736,70 @@ impl Env {
         contract_id
     }
 
-    /// Returns the recorded top-level `require_auth` or `require_auth_for_args`
-    /// calls that have happened during the last contract invocation.    
+    /// Set authorizations in the environment which will be consumed by
+    /// contracts when they invoke [`require_auth`] or [`require_auth_for_args`]
+    /// functions.
+    ///
+    /// To mock auth for testing, use [`mock_all_auths`]. If mocking of auths is
+    /// enabled, calling [`set_auths`] disables any mocking.
+    pub fn set_auths(&self, auths: &[ContractAuth]) {
+        self.env_impl
+            .set_authorization_entries(auths.to_vec())
+            .unwrap();
+    }
+
+    /// Mock all calls to the [`require_auth`] and [`require_auth_for_args`]
+    /// functions in invoked contracts, having them succeed as if authorization
+    /// was provided.
+    ///
+    /// When mocking is enabled, if the [`Address`] being authorized is the
+    /// address of a contract, that contract's `__check_auth` function will not
+    /// be called, and the contract does not need to exist or be registered in
+    /// the test.
+    ///
+    /// When mocking is enabled, if the [`Address`] being authorized is the
+    /// address of an account, the account does not need to exist.
+    ///
+    /// To disable mocking, see [`set_auths`].
+    ///
+    /// To access a list of auths that have occurred, see [`auths`].
+    ///
+    /// It is not currently possible to mock a subset of auths.
+    ///
+    /// ### Examples
+    /// ```
+    /// use soroban_sdk::{contractimpl, Env, Address, testutils::Address as _};
+    ///
+    /// pub struct HelloContract;
+    ///
+    /// #[contractimpl]
+    /// impl HelloContract {
+    ///     pub fn hello(env: Env, from: Address) {
+    ///         from.require_auth();
+    ///         // TODO
+    ///     }
+    /// }
+    ///
+    /// #[test]
+    /// fn test() {
+    /// # }
+    /// # fn main() {
+    ///     let env = Env::default();
+    ///     let contract_id = env.register_contract(None, HelloContract);
+    ///
+    ///     env.mock_all_auths();
+    ///
+    ///     let client = HelloContractClient::new(&env, &contract_id);
+    ///     let addr = Address::random(&env);
+    ///     client.hello(&addr);
+    /// }
+    /// ```
+    pub fn mock_all_auths(&self) {
+        self.env_impl.switch_to_recording_auth();
+    }
+
+    /// Returns a list of authorizations that were seen during the last contract
```

Comment:
These are still 'top' authorizations, but the comment doesn't really refer to that. While the previous definition probably had some room to improve the clarity, it contained information about what the 'top' means. That's important as e.g. internal token transfer calls won't be visible here and this is probably the most frequent scenario for the auth trees.

Reply:
Thanks for identifying this gap in the docs. I'd like to address the gap by removing top-level auths from the SDK for this function, to simplify the concepts a developer has to interact with. Opened https://github.com/stellar/rs-soroban-env/pull/787 and updated this pull request to use it instead. This PR is blocked on that env change.